// n8n/dist/src/PublicApi/v1/handlers/executions/executions.handler.js
"use strict";
const n8n_core_1 = require("n8n-core");
const executions_service_1 = require("./executions.service");
const __1 = require("../../../..");
const global_middleware_1 = require("../../shared/middlewares/global.middleware");
const workflows_service_1 = require("../workflows/workflows.service");
const pagination_service_1 = require("../../shared/services/pagination.service");
const InternalHooksManager_1 = require("../../../../InternalHooksManager");
module.exports = {
    deleteExecution: [
        (0, global_middleware_1.authorize)(['owner', 'member']),
        async (req, res) => {
            const sharedWorkflowsIds = await (0, workflows_service_1.getSharedWorkflowIds)(req.user);
            if (!sharedWorkflowsIds.length) {
                return res.status(404).json({ message: 'Not Found' });
            }
            const { id } = req.params;
            const execution = await (0, executions_service_1.getExecutionInWorkflows)(id, sharedWorkflowsIds, false);
            if (!execution) {
                return res.status(404).json({ message: 'Not Found' });
            }
            await n8n_core_1.BinaryDataManager.getInstance().deleteBinaryDataByExecutionId(execution.id.toString());
            await (0, executions_service_1.deleteExecution)(execution);
            execution.id = id;
            return res.json(execution);
        },
    ],
    getExecution: [
        (0, global_middleware_1.authorize)(['owner', 'member']),
        async (req, res) => {
            const sharedWorkflowsIds = await (0, workflows_service_1.getSharedWorkflowIds)(req.user);
            if (!sharedWorkflowsIds.length) {
                return res.status(404).json({ message: 'Not Found' });
            }
            const { id } = req.params;
            const { includeData = false } = req.query;
            const execution = await (0, executions_service_1.getExecutionInWorkflows)(id, sharedWorkflowsIds, includeData);
            if (!execution) {
                return res.status(404).json({ message: 'Not Found' });
            }
            void InternalHooksManager_1.InternalHooksManager.getInstance().onUserRetrievedExecution({
                user_id: req.user.id,
                public_api: true,
            });
            return res.json(execution);
        },
    ],
    getExecutions: [
        (0, global_middleware_1.authorize)(['owner', 'member']),
        global_middleware_1.validCursor,
        async (req, res) => {
            const { lastId = undefined, limit = 100, status = undefined, includeData = false, workflowId = undefined, } = req.query;
            const sharedWorkflowsIds = await (0, workflows_service_1.getSharedWorkflowIds)(req.user);
            if (!sharedWorkflowsIds.length) {
                return res.status(200).json({ data: [], nextCursor: null });
            }
            const runningExecutionsIds = __1.ActiveExecutions.getInstance()
                .getActiveExecutions()
                .map(({ id }) => Number(id));
            const filters = {
                status,
                limit,
                lastId,
                includeData,
                ...(workflowId && { workflowIds: [workflowId] }),
                excludedExecutionsIds: runningExecutionsIds,
            };
            const executions = await (0, executions_service_1.getExecutions)(filters);
            const newLastId = !executions.length ? 0 : executions.slice(-1)[0].id;
            filters.lastId = newLastId;
            const count = await (0, executions_service_1.getExecutionsCount)(filters);
            void InternalHooksManager_1.InternalHooksManager.getInstance().onUserRetrievedAllExecutions({
                user_id: req.user.id,
                public_api: true,
            });
            return res.json({
                data: executions,
                nextCursor: (0, pagination_service_1.encodeNextCursor)({
                    lastId: newLastId,
                    limit,
                    numberOfNextRecords: count,
                }),
            });
        },
    ],
};
