// n8n-nodes-base/dist/credentials/MattermostApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MattermostApi = void 0;
class MattermostApi {
    constructor() {
        this.name = 'mattermostApi';
        this.displayName = 'Mattermost API';
        this.documentationUrl = 'mattermost';
        this.properties = [
            {
                displayName: 'Access Token',
                name: 'accessToken',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
            {
                displayName: 'Base URL',
                name: 'baseUrl',
                type: 'string',
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.accessToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials.baseUrl}}/api/v4',
                url: '/users',
            },
        };
    }
}
exports.MattermostApi = MattermostApi;
