// n8n-nodes-base/dist/nodes/ExecuteWorkflowTrigger/ExecuteWorkflowTrigger.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExecuteWorkflowTrigger = void 0;
class ExecuteWorkflowTrigger {
    constructor() {
        this.description = {
            displayName: 'Execute Workflow Trigger',
            name: 'executeWorkflowTrigger',
            icon: 'fa:sign-out-alt',
            group: ['trigger'],
            version: 1,
            description: 'Runs the flow when called by the Execute Workflow node from a different workflow',
            maxNodes: 1,
            defaults: {
                name: 'When Called By Another Workflow',
                color: '#ff6d5a',
            },
            inputs: [],
            outputs: ['main'],
            properties: [
                {
                    displayName: 'When an ‘execute workflow’ node calls this workflow, the execution starts here. Any data passed into the \'execute workflow\' node will be output by this node.',
                    name: 'notice',
                    type: 'notice',
                    default: '',
                },
            ],
        };
    }
    execute() {
        const items = this.getInputData();
        return this.prepareOutputData(items);
    }
}
exports.ExecuteWorkflowTrigger = ExecuteWorkflowTrigger;
