// n8n/dist/src/PublicApi/v1/handlers/workflows/workflows.handler.js
"use strict";
const typeorm_1 = require("typeorm");
const __1 = require("../../../..");
const config = require("../../../../../config");
const WorkflowEntity_1 = require("../../../../databases/entities/WorkflowEntity");
const InternalHooksManager_1 = require("../../../../InternalHooksManager");
const Server_1 = require("../../../../Server");
const WorkflowHelpers_1 = require("../../../../WorkflowHelpers");
const global_middleware_1 = require("../../shared/middlewares/global.middleware");
const pagination_service_1 = require("../../shared/services/pagination.service");
const users_service_1 = require("../users/users.service");
const workflows_service_1 = require("./workflows.service");
module.exports = {
    createWorkflow: [
        (0, global_middleware_1.authorize)(['owner', 'member']),
        async (req, res) => {
            const workflow = req.body;
            workflow.active = false;
            if (!(0, workflows_service_1.hasStartNode)(workflow)) {
                workflow.nodes.push((0, workflows_service_1.getStartNode)());
            }
            await (0, WorkflowHelpers_1.replaceInvalidCredentials)(workflow);
            (0, WorkflowHelpers_1.addNodeIds)(workflow);
            const role = await (0, users_service_1.getWorkflowOwnerRole)();
            const createdWorkflow = await (0, workflows_service_1.createWorkflow)(workflow, req.user, role);
            await Server_1.externalHooks.run('workflow.afterCreate', [createdWorkflow]);
            void InternalHooksManager_1.InternalHooksManager.getInstance().onWorkflowCreated(req.user.id, createdWorkflow, true);
            return res.json(createdWorkflow);
        },
    ],
    deleteWorkflow: [
        (0, global_middleware_1.authorize)(['owner', 'member']),
        async (req, res) => {
            const { id } = req.params;
            const sharedWorkflow = await (0, workflows_service_1.getSharedWorkflow)(req.user, id.toString());
            if (!sharedWorkflow) {
                return res.status(404).json({ message: 'Not Found' });
            }
            if (sharedWorkflow.workflow.active) {
                await __1.ActiveWorkflowRunner.getInstance().remove(id.toString());
            }
            await __1.Db.collections.Workflow.delete(id);
            void InternalHooksManager_1.InternalHooksManager.getInstance().onWorkflowDeleted(req.user.id, id.toString(), true);
            await Server_1.externalHooks.run('workflow.afterDelete', [id.toString()]);
            return res.json(sharedWorkflow.workflow);
        },
    ],
    getWorkflow: [
        (0, global_middleware_1.authorize)(['owner', 'member']),
        async (req, res) => {
            const { id } = req.params;
            const sharedWorkflow = await (0, workflows_service_1.getSharedWorkflow)(req.user, id.toString());
            if (!sharedWorkflow) {
                return res.status(404).json({ message: 'Not Found' });
            }
            void InternalHooksManager_1.InternalHooksManager.getInstance().onUserRetrievedWorkflow({
                user_id: req.user.id,
                public_api: true,
            });
            return res.json(sharedWorkflow.workflow);
        },
    ],
    getWorkflows: [
        (0, global_middleware_1.authorize)(['owner', 'member']),
        global_middleware_1.validCursor,
        async (req, res) => {
            const { offset = 0, limit = 100, active = undefined, tags = undefined } = req.query;
            let workflows;
            let count;
            const query = {
                skip: offset,
                take: limit,
                where: {
                    ...(active !== undefined && { active }),
                },
                ...(!config.getEnv('workflowTagsDisabled') && { relations: ['tags'] }),
            };
            if ((0, users_service_1.isInstanceOwner)(req.user)) {
                if (tags) {
                    const workflowIds = await (0, workflows_service_1.getWorkflowIdsViaTags)((0, workflows_service_1.parseTagNames)(tags));
                    Object.assign(query.where, { id: (0, typeorm_1.In)(workflowIds) });
                }
                workflows = await (0, workflows_service_1.getWorkflows)(query);
                count = await (0, workflows_service_1.getWorkflowsCount)(query);
            }
            else {
                const options = {};
                if (tags) {
                    options.workflowIds = await (0, workflows_service_1.getWorkflowIdsViaTags)((0, workflows_service_1.parseTagNames)(tags));
                }
                const sharedWorkflows = await (0, workflows_service_1.getSharedWorkflows)(req.user, options);
                if (!sharedWorkflows.length) {
                    return res.status(200).json({
                        data: [],
                        nextCursor: null,
                    });
                }
                const workflowsIds = sharedWorkflows.map((shareWorkflow) => shareWorkflow.workflowId);
                Object.assign(query.where, { id: (0, typeorm_1.In)(workflowsIds) });
                workflows = await (0, workflows_service_1.getWorkflows)(query);
                count = await (0, workflows_service_1.getWorkflowsCount)(query);
            }
            void InternalHooksManager_1.InternalHooksManager.getInstance().onUserRetrievedAllWorkflows({
                user_id: req.user.id,
                public_api: true,
            });
            return res.json({
                data: workflows,
                nextCursor: (0, pagination_service_1.encodeNextCursor)({
                    offset,
                    limit,
                    numberOfTotalRecords: count,
                }),
            });
        },
    ],
    updateWorkflow: [
        (0, global_middleware_1.authorize)(['owner', 'member']),
        async (req, res) => {
            const { id } = req.params;
            const updateData = new WorkflowEntity_1.WorkflowEntity();
            Object.assign(updateData, req.body);
            const sharedWorkflow = await (0, workflows_service_1.getSharedWorkflow)(req.user, id.toString());
            if (!sharedWorkflow) {
                return res.status(404).json({ message: 'Not Found' });
            }
            if (!(0, workflows_service_1.hasStartNode)(updateData)) {
                updateData.nodes.push((0, workflows_service_1.getStartNode)());
            }
            await (0, WorkflowHelpers_1.replaceInvalidCredentials)(updateData);
            (0, WorkflowHelpers_1.addNodeIds)(updateData);
            const workflowRunner = __1.ActiveWorkflowRunner.getInstance();
            if (sharedWorkflow.workflow.active) {
                await workflowRunner.remove(id.toString());
            }
            try {
                await (0, workflows_service_1.updateWorkflow)(sharedWorkflow.workflowId, updateData);
            }
            catch (error) {
                if (error instanceof Error) {
                    return res.status(400).json({ message: error.message });
                }
            }
            if (sharedWorkflow.workflow.active) {
                try {
                    await workflowRunner.add(sharedWorkflow.workflowId.toString(), 'update');
                }
                catch (error) {
                    if (error instanceof Error) {
                        return res.status(400).json({ message: error.message });
                    }
                }
            }
            const updatedWorkflow = await (0, workflows_service_1.getWorkflowById)(sharedWorkflow.workflowId);
            await Server_1.externalHooks.run('workflow.afterUpdate', [updateData]);
            void InternalHooksManager_1.InternalHooksManager.getInstance().onWorkflowSaved(req.user.id, updateData, true);
            return res.json(updatedWorkflow);
        },
    ],
    activateWorkflow: [
        (0, global_middleware_1.authorize)(['owner', 'member']),
        async (req, res) => {
            const { id } = req.params;
            const sharedWorkflow = await (0, workflows_service_1.getSharedWorkflow)(req.user, id.toString());
            if (!sharedWorkflow) {
                return res.status(404).json({ message: 'Not Found' });
            }
            if (!sharedWorkflow.workflow.active) {
                try {
                    await __1.ActiveWorkflowRunner.getInstance().add(sharedWorkflow.workflowId.toString(), 'activate');
                }
                catch (error) {
                    if (error instanceof Error) {
                        return res.status(400).json({ message: error.message });
                    }
                }
                await (0, workflows_service_1.setWorkflowAsActive)(sharedWorkflow.workflow);
                sharedWorkflow.workflow.active = true;
                return res.json(sharedWorkflow.workflow);
            }
            return res.json(sharedWorkflow.workflow);
        },
    ],
    deactivateWorkflow: [
        (0, global_middleware_1.authorize)(['owner', 'member']),
        async (req, res) => {
            const { id } = req.params;
            const sharedWorkflow = await (0, workflows_service_1.getSharedWorkflow)(req.user, id.toString());
            if (!sharedWorkflow) {
                return res.status(404).json({ message: 'Not Found' });
            }
            const workflowRunner = __1.ActiveWorkflowRunner.getInstance();
            if (sharedWorkflow.workflow.active) {
                await workflowRunner.remove(sharedWorkflow.workflowId.toString());
                await (0, workflows_service_1.setWorkflowAsInactive)(sharedWorkflow.workflow);
                sharedWorkflow.workflow.active = false;
                return res.json(sharedWorkflow.workflow);
            }
            return res.json(sharedWorkflow.workflow);
        },
    ],
};
