// n8n-nodes-base/dist/nodes/Aws/Rekognition/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.keysTPascalCase = exports.awsApiRequestSOAPAllItems = exports.awsApiRequestSOAP = exports.awsApiRequestREST = exports.awsApiRequest = void 0;
const lodash_1 = require("lodash");
const xml2js_1 = require("xml2js");
const n8n_workflow_1 = require("n8n-workflow");
const change_case_1 = require("change-case");
async function awsApiRequest(service, method, path, body, query = {}, headers, option = {}, region) {
    const credentials = await this.getCredentials('aws');
    const requestOptions = {
        qs: {
            service,
            path,
        },
        method,
        body,
        url: '',
        headers,
        region: credentials === null || credentials === void 0 ? void 0 : credentials.region,
    };
    if (Object.keys(option).length !== 0) {
        Object.assign(requestOptions, option);
    }
    try {
        return await this.helpers.requestWithAuthentication.call(this, 'aws', requestOptions);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.awsApiRequest = awsApiRequest;
async function awsApiRequestREST(service, method, path, body, query = {}, headers, options = {}, region) {
    const response = await awsApiRequest.call(this, service, method, path, body, query, headers, options, region);
    try {
        return JSON.parse(response);
    }
    catch (error) {
        return response;
    }
}
exports.awsApiRequestREST = awsApiRequestREST;
async function awsApiRequestSOAP(service, method, path, body, query = {}, headers, option = {}, region) {
    const response = await awsApiRequest.call(this, service, method, path, body, query, headers, option, region);
    try {
        return await new Promise((resolve, reject) => {
            (0, xml2js_1.parseString)(response, { explicitArray: false }, (err, data) => {
                if (err) {
                    return reject(err);
                }
                resolve(data);
            });
        });
    }
    catch (error) {
        return error;
    }
}
exports.awsApiRequestSOAP = awsApiRequestSOAP;
async function awsApiRequestSOAPAllItems(propertyName, service, method, path, body, query = {}, headers = {}, option = {}, region) {
    const returnData = [];
    let responseData;
    do {
        responseData = await awsApiRequestSOAP.call(this, service, method, path, body, query, headers, option, region);
        if ((0, lodash_1.get)(responseData, `${propertyName.split('.')[0]}.NextContinuationToken`)) {
            query['continuation-token'] = (0, lodash_1.get)(responseData, `${propertyName.split('.')[0]}.NextContinuationToken`);
        }
        if ((0, lodash_1.get)(responseData, propertyName)) {
            if (Array.isArray((0, lodash_1.get)(responseData, propertyName))) {
                returnData.push.apply(returnData, (0, lodash_1.get)(responseData, propertyName));
            }
            else {
                returnData.push((0, lodash_1.get)(responseData, propertyName));
            }
        }
        if (query.limit && query.limit <= returnData.length) {
            return returnData;
        }
    } while ((0, lodash_1.get)(responseData, `${propertyName.split('.')[0]}.IsTruncated`) !== undefined &&
        (0, lodash_1.get)(responseData, `${propertyName.split('.')[0]}.IsTruncated`) !== 'false');
    return returnData;
}
exports.awsApiRequestSOAPAllItems = awsApiRequestSOAPAllItems;
function queryToString(params) {
    return Object.keys(params)
        .map((key) => key + '=' + params[key])
        .join('&');
}
function keysTPascalCase(object) {
    const data = {};
    for (const key of Object.keys(object)) {
        data[(0, change_case_1.pascalCase)(key)] = object[key];
    }
    return data;
}
exports.keysTPascalCase = keysTPascalCase;
