// n8n/dist/commands/BaseCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseCommand = void 0;
const core_1 = require("@oclif/core");
const n8n_workflow_1 = require("n8n-workflow");
const Logger_1 = require("../src/Logger");
const User_1 = require("../src/databases/entities/User");
const src_1 = require("../src");
class BaseCommand extends core_1.Command {
    constructor() {
        super(...arguments);
        this.defaultUserProps = {
            firstName: null,
            lastName: null,
            email: null,
            password: null,
            resetPasswordToken: null,
        };
    }
    async init() {
        this.logger = (0, Logger_1.getLogger)();
        n8n_workflow_1.LoggerProxy.init(this.logger);
        await src_1.Db.init();
    }
    async finally() {
        if (process.env.NODE_ENV === 'test')
            return;
        this.exit();
    }
    async getInstanceOwner() {
        const globalRole = await src_1.Db.collections.Role.findOneOrFail({
            name: 'owner',
            scope: 'global',
        });
        const owner = await src_1.Db.collections.User.findOne({ globalRole });
        if (owner)
            return owner;
        const user = new User_1.User();
        Object.assign(user, { ...this.defaultUserProps, globalRole });
        await src_1.Db.collections.User.save(user);
        return src_1.Db.collections.User.findOneOrFail({ globalRole });
    }
}
exports.BaseCommand = BaseCommand;
