// n8n/dist/src/databases/entities/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.entities = void 0;
const CredentialsEntity_1 = require("./CredentialsEntity");
const ExecutionEntity_1 = require("./ExecutionEntity");
const WorkflowEntity_1 = require("./WorkflowEntity");
const WebhookEntity_1 = require("./WebhookEntity");
const TagEntity_1 = require("./TagEntity");
const User_1 = require("./User");
const Role_1 = require("./Role");
const Settings_1 = require("./Settings");
const SharedWorkflow_1 = require("./SharedWorkflow");
const SharedCredentials_1 = require("./SharedCredentials");
const InstalledPackages_1 = require("./InstalledPackages");
const InstalledNodes_1 = require("./InstalledNodes");
exports.entities = {
    CredentialsEntity: CredentialsEntity_1.CredentialsEntity,
    ExecutionEntity: ExecutionEntity_1.ExecutionEntity,
    WorkflowEntity: WorkflowEntity_1.WorkflowEntity,
    WebhookEntity: WebhookEntity_1.WebhookEntity,
    TagEntity: TagEntity_1.TagEntity,
    User: User_1.User,
    Role: Role_1.Role,
    Settings: Settings_1.Settings,
    SharedWorkflow: SharedWorkflow_1.SharedWorkflow,
    SharedCredentials: SharedCredentials_1.SharedCredentials,
    InstalledPackages: InstalledPackages_1.InstalledPackages,
    InstalledNodes: InstalledNodes_1.InstalledNodes,
};
