// n8n/dist/src/databases/migrations/sqlite/1659888469333-AddJsonKeyPinData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addJsonKeyToPinDataColumn = exports.AddJsonKeyPinData1659888469333 = void 0;
const migrationHelpers_1 = require("../../utils/migrationHelpers");
const migrations_types_1 = require("../../utils/migrations.types");
class AddJsonKeyPinData1659888469333 {
    constructor() {
        this.name = 'AddJsonKeyPinData1659888469333';
    }
    async up(queryRunner) {
        (0, migrationHelpers_1.logMigrationStart)(this.name);
        const workflowTable = `${(0, migrationHelpers_1.getTablePrefix)()}workflow_entity`;
        const PINDATA_SELECT_QUERY = `
			SELECT id, pinData
			FROM "${workflowTable}"
			WHERE pinData IS NOT NULL;
		`;
        const PINDATA_UPDATE_STATEMENT = `
			UPDATE "${workflowTable}"
			SET "pinData" = :pinData
			WHERE id = :id;
		`;
        await (0, migrationHelpers_1.runInBatches)(queryRunner, PINDATA_SELECT_QUERY, (0, exports.addJsonKeyToPinDataColumn)(queryRunner, PINDATA_UPDATE_STATEMENT));
        (0, migrationHelpers_1.logMigrationEnd)(this.name);
    }
    async down() {
    }
}
exports.AddJsonKeyPinData1659888469333 = AddJsonKeyPinData1659888469333;
const addJsonKeyToPinDataColumn = (queryRunner, updateStatement) => async (fetchedWorkflows) => {
    makeUpdateParams(fetchedWorkflows).forEach((param) => {
        const params = {
            pinData: param.pinData,
            id: param.id,
        };
        const [escapedStatement, escapedParams] = (0, migrationHelpers_1.escapeQuery)(queryRunner, updateStatement, params);
        queryRunner.query(escapedStatement, escapedParams);
    });
};
exports.addJsonKeyToPinDataColumn = addJsonKeyToPinDataColumn;
function makeUpdateParams(fetchedWorkflows) {
    return fetchedWorkflows.reduce((updateParams, { id, pinData: rawPinData }) => {
        let pinDataPerWorkflow;
        if (typeof rawPinData === 'string') {
            try {
                pinDataPerWorkflow = JSON.parse(rawPinData);
            }
            catch (_) {
                pinDataPerWorkflow = {};
            }
        }
        else {
            pinDataPerWorkflow = rawPinData;
        }
        const newPinDataPerWorkflow = Object.keys(pinDataPerWorkflow).reduce((newPinDataPerWorkflow, nodeName) => {
            let pinDataPerNode = pinDataPerWorkflow[nodeName];
            if (!Array.isArray(pinDataPerNode)) {
                pinDataPerNode = [pinDataPerNode];
            }
            if (pinDataPerNode.every((item) => item.json))
                return newPinDataPerWorkflow;
            newPinDataPerWorkflow[nodeName] = pinDataPerNode.map((item) => (0, migrations_types_1.isJsonKeyObject)(item) ? item : { json: item });
            return newPinDataPerWorkflow;
        }, {});
        if (Object.keys(newPinDataPerWorkflow).length > 0) {
            updateParams.push({ id, pinData: JSON.stringify(newPinDataPerWorkflow) });
        }
        return updateParams;
    }, []);
}
