// n8n-nodes-base/dist/nodes/Todoist/Service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OperationType = exports.TodoistService = void 0;
const OperationHandler_1 = require("./OperationHandler");
class TodoistService {
    constructor() {
        this.handlers = {
            create: new OperationHandler_1.CreateHandler(),
            close: new OperationHandler_1.CloseHandler(),
            delete: new OperationHandler_1.DeleteHandler(),
            get: new OperationHandler_1.GetHandler(),
            getAll: new OperationHandler_1.GetAllHandler(),
            reopen: new OperationHandler_1.ReopenHandler(),
            update: new OperationHandler_1.UpdateHandler(),
            move: new OperationHandler_1.MoveHandler(),
            sync: new OperationHandler_1.SyncHandler(),
        };
    }
    async execute(ctx, operation, itemIndex) {
        return this.handlers[operation].handleOperation(ctx, itemIndex);
    }
}
exports.TodoistService = TodoistService;
var OperationType;
(function (OperationType) {
    OperationType["create"] = "create";
    OperationType["close"] = "close";
    OperationType["delete"] = "delete";
    OperationType["get"] = "get";
    OperationType["getAll"] = "getAll";
    OperationType["reopen"] = "reopen";
    OperationType["update"] = "update";
    OperationType["move"] = "move";
    OperationType["sync"] = "sync";
})(OperationType = exports.OperationType || (exports.OperationType = {}));
