// n8n-nodes-base/dist/nodes/Twilio/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.escapeXml = exports.twilioApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function twilioApiRequest(method, endpoint, body, query) {
    const credentials = (await this.getCredentials('twilioApi'));
    if (query === undefined) {
        query = {};
    }
    const options = {
        method,
        form: body,
        qs: query,
        uri: `https://api.twilio.com/2010-04-01/Accounts/${credentials.accountSid}${endpoint}`,
        json: true,
    };
    try {
        return await this.helpers.requestWithAuthentication.call(this, 'twilioApi', options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.twilioApiRequest = twilioApiRequest;
const XML_CHAR_MAP = {
    '<': '&lt;',
    '>': '&gt;',
    '&': '&amp;',
    '"': '&quot;',
    "'": '&apos;',
};
function escapeXml(str) {
    return str.replace(/[<>&"']/g, (ch) => {
        return XML_CHAR_MAP[ch];
    });
}
exports.escapeXml = escapeXml;
