// n8n/dist/src/UserManagement/routes/me.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.meNamespace = void 0;
const validator_1 = __importDefault(require("validator"));
const n8n_workflow_1 = require("n8n-workflow");
const crypto_1 = require("crypto");
const __1 = require("../..");
const jwt_1 = require("../auth/jwt");
const UserManagementHelper_1 = require("../UserManagementHelper");
const GenericHelpers_1 = require("../../GenericHelpers");
const User_1 = require("../../databases/entities/User");
function meNamespace() {
    this.app.get(`/${this.restEndpoint}/me`, __1.ResponseHelper.send(async (req) => {
        return (0, UserManagementHelper_1.sanitizeUser)(req.user);
    }));
    this.app.patch(`/${this.restEndpoint}/me`, __1.ResponseHelper.send(async (req, res) => {
        const { email } = req.body;
        if (!email) {
            n8n_workflow_1.LoggerProxy.debug('Request to update user email failed because of missing email in payload', {
                userId: req.user.id,
                payload: req.body,
            });
            throw new __1.ResponseHelper.ResponseError('Email is mandatory', undefined, 400);
        }
        if (!validator_1.default.isEmail(email)) {
            n8n_workflow_1.LoggerProxy.debug('Request to update user email failed because of invalid email in payload', {
                userId: req.user.id,
                invalidEmail: email,
            });
            throw new __1.ResponseHelper.ResponseError('Invalid email address', undefined, 400);
        }
        const { email: currentEmail } = req.user;
        const newUser = new User_1.User();
        Object.assign(newUser, req.user, req.body);
        await (0, GenericHelpers_1.validateEntity)(newUser);
        const user = await __1.Db.collections.User.save(newUser);
        n8n_workflow_1.LoggerProxy.info('User updated successfully', { userId: user.id });
        await (0, jwt_1.issueCookie)(res, user);
        const updatedkeys = Object.keys(req.body);
        void __1.InternalHooksManager.getInstance().onUserUpdate({
            user_id: req.user.id,
            fields_changed: updatedkeys,
        });
        await this.externalHooks.run('user.profile.update', [currentEmail, (0, UserManagementHelper_1.sanitizeUser)(user)]);
        return (0, UserManagementHelper_1.sanitizeUser)(user);
    }));
    this.app.patch(`/${this.restEndpoint}/me/password`, __1.ResponseHelper.send(async (req, res) => {
        const { currentPassword, newPassword } = req.body;
        if (typeof currentPassword !== 'string' || typeof newPassword !== 'string') {
            throw new __1.ResponseHelper.ResponseError('Invalid payload.', undefined, 400);
        }
        if (!req.user.password) {
            throw new __1.ResponseHelper.ResponseError('Requesting user not set up.');
        }
        const isCurrentPwCorrect = await (0, UserManagementHelper_1.compareHash)(currentPassword, req.user.password);
        if (!isCurrentPwCorrect) {
            throw new __1.ResponseHelper.ResponseError('Provided current password is incorrect.', undefined, 400);
        }
        const validPassword = (0, UserManagementHelper_1.validatePassword)(newPassword);
        req.user.password = await (0, UserManagementHelper_1.hashPassword)(validPassword);
        const user = await __1.Db.collections.User.save(req.user);
        n8n_workflow_1.LoggerProxy.info('Password updated successfully', { userId: user.id });
        await (0, jwt_1.issueCookie)(res, user);
        void __1.InternalHooksManager.getInstance().onUserUpdate({
            user_id: req.user.id,
            fields_changed: ['password'],
        });
        await this.externalHooks.run('user.password.update', [user.email, req.user.password]);
        return { success: true };
    }));
    this.app.post(`/${this.restEndpoint}/me/survey`, __1.ResponseHelper.send(async (req) => {
        const { body: personalizationAnswers } = req;
        if (!personalizationAnswers) {
            n8n_workflow_1.LoggerProxy.debug('Request to store user personalization survey failed because of empty payload', {
                userId: req.user.id,
            });
            throw new __1.ResponseHelper.ResponseError('Personalization answers are mandatory', undefined, 400);
        }
        await __1.Db.collections.User.save({
            id: req.user.id,
            personalizationAnswers,
        });
        n8n_workflow_1.LoggerProxy.info('User survey updated successfully', { userId: req.user.id });
        void __1.InternalHooksManager.getInstance().onPersonalizationSurveySubmitted(req.user.id, personalizationAnswers);
        return { success: true };
    }));
    this.app.post(`/${this.restEndpoint}/me/api-key`, __1.ResponseHelper.send(async (req) => {
        const apiKey = `n8n_api_${(0, crypto_1.randomBytes)(40).toString('hex')}`;
        await __1.Db.collections.User.update(req.user.id, {
            apiKey,
        });
        const telemetryData = {
            user_id: req.user.id,
            public_api: false,
        };
        void __1.InternalHooksManager.getInstance().onApiKeyCreated(telemetryData);
        return { apiKey };
    }));
    this.app.delete(`/${this.restEndpoint}/me/api-key`, __1.ResponseHelper.send(async (req) => {
        await __1.Db.collections.User.update(req.user.id, {
            apiKey: null,
        });
        const telemetryData = {
            user_id: req.user.id,
            public_api: false,
        };
        void __1.InternalHooksManager.getInstance().onApiKeyDeleted(telemetryData);
        return { success: true };
    }));
    this.app.get(`/${this.restEndpoint}/me/api-key`, __1.ResponseHelper.send(async (req) => {
        return { apiKey: req.user.apiKey };
    }));
}
exports.meNamespace = meNamespace;
